SEQUANA
############

.. image:: https://badge.fury.io/py/sequana.svg
    :target: https://pypi.python.org/pypi/sequana

.. image:: https://travis-ci.org/sequana/sequana.svg?branch=master
    :target: https://travis-ci.org/sequana/sequana

.. image:: https://coveralls.io/repos/github/sequana/sequana/badge.svg?branch=master
    :target: https://coveralls.io/github/sequana/sequana?branch=master 

.. image:: http://readthedocs.org/projects/sequana/badge/?version=master
    :target: http://sequana.readthedocs.org/en/latest/?badge=master
    :alt: Documentation Status

:Python version: 3.5 (and 2.7 without snakemake)
:Online documentation: `On readthedocs <http://sequana.readthedocs.org/>`_
:Issues and bug reports: `On github <https://github.com/sequana/sequana/issues>`_
:How to cite: Desvillechabrol D, Bouchier C, Cokelaer T and Kennedy S. Sequana: a set of
    flexible genomic pipelines for processing and reporting NGS analysis [v1; no peer reviewed]. 
    `F1000Research 2016, 5:1767 <http://f1000research.com/posters/5-1767>`_ (poster) (doi:
    10.7490/f1000research.1112656.1)


**Sequana** includes a set of pipelines related to NGS (new generation sequencing) including quality control, variant calling, coverage, taxonomy...

**Please see the** `documentation <http://sequana.readthedocs.org>`_ for an
up-to-date status and documentation.


Docker images (experimental):
-------------------------------

.. image:: https://images.microbadger.com/badges/image/cokelaer/sequana.svg
    :target: https://microbadger.com/images/cokelaer/sequana
    :alt: Docker info

.. image:: https://images.microbadger.com/badges/version/cokelaer/sequana.svg
    :target: https://microbadger.com/images/cokelaer/sequana
    :alt: Docker hub




