FAQS
======

Conda related
---------------

Create a conda environment on IP cluster::

    module load conda
    conda create --name py35 python=3.5
    source condaenvs/py35/bin/activate py35

add channel from where to download packages::

    conda config --add channels r bioconda
    conda install sequana


What are the dependencies
-----------------------------

There are two kind of dependencies. First, the Python libraries such as
matplotlib or Pandas. Second, the external tools such as BWA (alignment) or
Kraken (taxonomy). The first kind of tools can be installed using Anaconda and the
default conda channel. For instance::

    conda install pandas

The second kind of tools can also be installed using another conda channel
called **bioconda**. For instance::

    conda install bwa

The full list of dependencies will be maintained in the :ref:`installation`
section but those two lines should be sufficient to install most of the
dependencies with **conda**::

    conda install --file https://raw.githubusercontent.com/sequana/sequana/master/requirements.txt
    conda install --file https://raw.githubusercontent.com/sequana/sequana/master/requirements_pipelines.txt

Additional tools such as prokka, busco, canu and future heavy software will be
maintained in this specific requirements for now::

    conda install --file https://raw.githubusercontent.com/sequana/sequana/master/requirements_pipelines_extra.txt



Installation issues
-----------------------


As explained in the previous section, most of the dependencies can be installed
via Conda. If not, pip is recommended. Yet there are still a few dependencies
that needs manual installation. 

quast
~~~~~~~~~

http://quast.bioinf.spbau.ru/manual.html#sec1

::

    wget https://downloads.sourceforge.net/project/quast/quast-4.2.tar.gz
    tar -xzf quast-4.2.tar.gz
    cd quast-4.2

Alternatively, get the source code from their GitHub (takes a while)::

    git clone https://github.com/ablab/quast
    cd quast
    python setup.py install

graphviz
~~~~~~~~~~~~~~~~~~

graphviz provides an executable called **dot**. If you type **dot** in a shell
and get this error message::

    Warning: Could not load
    ...lib/graphviz/libgvplugin_gd.so.6" - file not found

This may be solved by re-installation graphviz using the main anaconda channel
(instead of bioconda)::

    conda install --override-channels -c anaconda graphviz=2.38.0 

:Update April 2017: replace anaconda with conda-forge


matplotlib
~~~~~~~~~~~~~~~~~

If you get errors related to the X connection, you may need to change the
backend of matplotlib. To do so, go in your home directory and in this directory

    /home/user/.config/matplotlib ,

add a file called **matplotlibrc** with the following content::

    backend: Agg

Save, exit the shell, start a new shell.


pysam / samtools / bzip2
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

We have experienced few issues with pysam and samtools. Here are some solutions.


::

    from pysam.libchtslib import *
    ...ImportError: libhts.so.1: cannot open shared object file: No such file or directory


This may be solved by removing conda installation and using pip instead::

     conda remove pysam
     pip install pysam

Another error know for pysam version 0.11.2.2 raises this error::

    ImportError: libbz2.so.1.0: cannot open shared object file: No such file or
    directory

Downgrading to version 0.11.2.1 and upgrading to working version solves the problem::

    conda install pysam=0.11.2.1

but one reason was also related to the order of the channel in the .condarc
file. You may get bzip2 from the default channel and not from
conda-forge (reference: https://github.com/bioconda/bioconda-recipes/issues/5188)
::

    conda install --override-channels -c conda-forge bzip2



qt
~~~~~~~~~~~~~~~~~~
::

    from PyQt5.QtWebKitWidgets import QWebView
    ...ImportError: libQt5WebKitWidgets.so.5: cannot open shared object file: No such file or directory

This may be solved by re-installation qt using the main anaconda channel
(instead of bioconda)::

    conda install --override-channels -c anaconda qt


libselinux
~~~~~~~~~~~~~~~~~

If you get this error (using **conda install sequana**)::

    ImportError: libselinux.so.1: cannot open shared object file: No such file or directory

it looks like you need to install libselinux on your environment as reported 
`here <https://github.com/sequana/sequana/issues/438>`_.


Expected input format
----------------------------

Most of the pipelines and standalone expect FastQ files with the extension
**fastq.gz** meaning that files are gzipped.


Besides, the filename convention is as follows::

    PREFIX_R1_.fastq.gz

that is **_R1_** and **_R2_** indicates the paired or single-ended files and
the PREFIX is used to create directories or reports; it must be present.

.. versionadded:: 0.2
    more flexible tags are now possible in sequana pipelines and sequanix using
    e.g. _R[12] in the **input_readtag** in the configuration file of the
    pipelines.


Sequanix related
----------------------

For question related to Sequanix, we have a dedicated section in
:ref:`sequanix_faqs`.


QXcbConnection issue
----------------------
If you get this error::

    QXcbConnection: Could not connect to display localhost:10.0

this is an issue with your Qt backend. You need to change it to Agg.




Variant Calling pipeline
----------------------------

If snpeff fails with this type of errors::

    java.lang.RuntimeException: Error reading file 'null'
    java.lang.RuntimeException: Cannot find sequence for 'LN831026.gbk'

this may be because your genbank does not contain the sequences.

Another type of errors is that the sequence and genbank are not synchrone. We
would recommend to use the code here to download the Fasta and genbank:

http://sequana.readthedocs.io/en/master/tutorial.html#new-in-v0-10






Singularity
-----------------

If you use the singularity container and get this kind of error::

    singularity shell sequana-sequana-master.img
    ERROR  : Base home directory does not exist within the container: /pasteur
    ABORT  : Retval = 255

it means the container does not know about the Base home directory.

If you have sudo access, add the missing path as follows::

    sudo singularity shell --writable sequana-sequana-master.img
    mkdir /pasteur
    exit

If you do not have sudo permissions, copy the image on a computer where you have
such permission, use the same code as above and copy back the new image on the
computer where you had the issue. 

Finally, try to use the container again using this code::

    singularity shell sequana-sequana-master.img







